scalyr_agent/util.py

```diff
@@ -560,7 +574,7 @@ def rfc3339_to_datetime(string):
     return dt
 
 
-def rfc3339_to_nanoseconds_since_epoch(string):
+def rfc3339_to_nanoseconds_since_epoch(string, use_strptime=False):
```

Comment:
I take it you are using the benchmark test to cover functional testing as well?  Philosophically, I more learn towards also including a test case for this in the `util_tests.py` file mainly for consistency and discoverability.  If I modify this code, I would expect to find its test cases in `util_tests.py`.  I can understand though that it is duplication of effort / code if we went this way.  What are your thoughts?

Reply:
I can also add it in there.

In fact, I'm still deciding how to do it. Pretty much the only reason I added this flag is so we can benchmark and compare the versions (to ensure correctness).

Now that we know that the new version is faster and correct, we could simply remove that flag all togethers.

What's your take on this?

I'm personally fine either way. Perhaps for now I can also add test case to util_tests.py and in the near future, we can remove that flag all together and simplify the code.

Reply:
I've added additional round-trip correctness test cases - 3a4df848570ce885976786c16b0de1d56de25727.

After some more thinking, I think the best approach is to have three functions - default one which utilizes the fastest approach (aka default public one) and then one which uses regex and one which uses striptime.

The last two would be "private" and only exported and used for testing, benchmarking and auditability purposes.

Reply:
@imron suggested another approach - https://github.com/freach/udatetime

Since it's written in C, it's likely that it's faster, but need to verify it since our current implementation doesn't support timezones and udatetime does so there is likely some overhead related to timezone parsing and handling.

In any case, I will go with "multiple functions" approach so we can easily benchmark multiple implementations and compare one to another.

Oh and since that library is posix only, we will still need a fallback for platform and situations where that library is not available (Windows and other platforms where a pre-build wheel / egg is not available).

For out of the box installation, we can try to bundle the pre-built posix egg, but we still need a fallback regardless since that egg is platform specific and may not work in a lot of situations.